### DISCUSSION TITLE
server-side rendering components

### DISCUSSION THREAD

**Discussion by KolyaDychkevych** [vercel/next.js#66955] [Help]
### Summary

i want to render components on server side 

I have created serverRegisterComponents with compoennts
and the createComponents function, which gets the components by type, then converts to html str and sends to the next step - its OK

The problem is that every component from the registry has fetch requests inside and I don't get data from those requests because the fetch requests are asynchronous and the server side rendering is synchronous.

maybe there needs to be a different way to render the components or how to wait for the requests to complete and then render the component

### Additional information

```js
const serverComponentsRegister: { [key: string]: any } = {
  product_list: ProductList,
  chart: Chart,
  tableComponent: TableComponent,
  list: List
};

const createComponent = async (type: string): Promise<string> => {
  try {
    const Component =  serverComponentsRegister[type];
   
    const element = React.createElement(Component);
    const htmlOutput = ReactDOMServer.renderToString(element);
    
    return htmlOutput;
  } catch (error) {
    throw new Error(`Error creating component: ${error.message}`);
  }
};

// API route
export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.method === 'POST') {
    const { type, properties } = req.body;
    try {
      const output = await createComponent(type, properties);
      
      res.status(200).json({html: output});
    } catch (error) {
      res.status(500).json({ error: error.message });
    }
  } else {
    res.status(405).json({ message: 'Only POST requests are allowed' });
  }
}
```

**Comment by icyJoseph** [comment]
Maybe you wanna use `https://react.dev/reference/react-dom/server/renderToNodeStream`, but be aware that:

- It is a deprecated API, and will be removed at some point
- This method will wait for all [Suspense boundaries](https://react.dev/reference/react/Suspense) to complete before returning any output.

Maybe implement something like, https://react.dev/reference/react-dom/server/renderToPipeableStream

At which point... you might as well just make a page component... I think in Pages Router, you could also peek into the body of POST request, in getServerSideProps, right? I do not remember... but definitely not in App Router T_T
